### PR TITLE
:bug: Return NoResourceMatchError when appropriate for backwards compatibility.

### PR DESCRIPTION
**Pull request by porridge**
Fixes: #2354

### CODE REVIEW

File: pkg/client/apiutil/errors.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	sort.Strings(subErrors)
+	return fmt.Sprintf("unable to retrieve the complete list of server APIs: %s", strings.Join(subErrors, ", "))
+}
+
+func (e *ErrResourceDiscoveryFailed) Unwrap() []error {
+	subErrors := []error{}
+	for gv, err := range *e {
+		if apierrors.IsNotFound(err) {
+			err = &meta.NoResourceMatchError{PartialResource: gv.WithResource("")}
```
Comment (alvaroaleman) [inline pkg/client/apiutil/errors.go:49]: This confuses me - In the initial issue description, @saschagrunert mentions that they expected to match through IsNotFound: https://github.com/kubernetes-sigs/controller-runtime/issues/2354#issue-1727232531 - Which should this be?

**Comment by porridge** [comment]
@alvaroaleman good point, let's see.

I created [a tiny repro program](https://github.com/porridge/controller-runtime/blob/check-on-v0.14.6/t/retval_test.go) to investigate the behaviour of an actual **client** that uses the mapper, and ran it on a few versions:
* pre-v0.15
   ```
    deleg.go:124: INFO ret-error-demo: Kind is present but object was not found kind="ReplicaSet"
    deleg.go:124: INFO ret-error-demo: Kind does not exist kind="Badger"
    deleg.go:124: INFO ret-error-demo: Kind does not exist kind="Gherkin"
   ```
* curent tip of `main` ([commit](https://github.com/porridge/controller-runtime/commit/89ae07ce10bad2f4d1c73f03df025c59438208b2)):
   ```
    deleg.go:127: I0911 12:22:01.453070] ret-error-demo: Kind is present but object was not found kind="ReplicaSet"
    deleg.go:127: I0911 12:22:01.453660] ret-error-demo: Kind does not exist kind="Badger"
    deleg.go:142: E0911 12:22:01.454161] ret-error-demo: Unexpected error err="failed to get API group resources: unable to retrieve the complete list of server APIs: inexistent.group.com/v1: the server could not find the requested resource"
   ```
* on top of this PR ([commit](https://github.com/porridge/controller-runtime/commit/fea1836291f7605b6e5d3a3b49cd6610527aa6a2)):
   ```
    deleg.go:127: I0911 12:24:06.089190] ret-error-demo: Kind is present but object was not found kind="ReplicaSet"
    deleg.go:127: I0911 12:24:06.089777] ret-error-demo: Kind does not exist kind="Badger"
    deleg.go:127: I0911 12:24:06.090234] ret-error-demo: Kind does not exist kind="Gherkin"
   ```

BTW, this also shows that the regression is only for the case of missing group(version) - error returned on a missing kind in an existing group+version is ok w.r.t. backwards compatibility. Unfortunately the missing group/version case is the most interesting one, as it's used to discover capabilities supported by the cluster.

My theory on why @saschagrunert mentioned `IsNotFound` is that it was a mistake or simply lack of precision. At least judging by the code he mentions, it contains a [single method that ORs together checks](https://github.com/kubernetes-sigs/security-profiles-operator/pull/1710/files#diff-fb01936f787b03e615f35a185799e91ec41896c1fcae0991db3a24137f21cf54R73-R74) done by `IsNoMatchError` and `IsNotFound`.

The conditional branches in the demo program I created mimics [actual the code in my operator](https://github.com/stackrox/stackrox/blob/de940b3d011d5bcb9a2519f916664ff4ce25588f/operator/pkg/securedcluster/scanner/auto_sense.go#L72-L81) which is careful to distinguish between errors that satisfy `IsNoMatchError` vs `IsNotFound`.

**Review by alvaroaleman — APPROVED** [review]
Thanks for fixing this!

**Pull request merged**